Make ToxAV stateless

GrayHatter opened the pull request:
See #40.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/toxcore/130)
<!-- Reviewable:end -->

iphydf commented:
Review status: 0 of 20 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[other/apidsl/toxav.in.h, line 34 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT7FOEn_7CC8jZVi-i:-KRT7FOFhFzUv0-M7I2Y:b6p9e0r) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/other/apidsl/toxav.in.h#L34)):*
> ```objc
> #endif
> 
> #define TOXAV_PAIR(TYPE1__, TYPE2__) struct { TYPE1__ first; TYPE2__ second; }
> ```

Why is this in the public API?

---

*[other/apidsl/toxav.in.h, line 221 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT7NPzgOczW4g0e7eR:-KRT7NPzgOczW4g0e7eS:b-pc7ae0) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/other/apidsl/toxav.in.h#L221)):*
> ```objc
>  * thread from tox_iterate.
>  */
> void iterate(void *userdata);
> ```

`any userdata` like in tox.in.h.

---

*[toxav/audio.c, line 99 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT7lEVOj5USXMPbwRq:-KRT7lEVOj5USXMPbwRr:b-97il5) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/toxav/audio.c#L99)):*
> ```C
>     ac->av = av;
>     ac->friend_number = friend_number;
>     ac->toxav_audio_frame_fxn = cb;
> ```

I very much dislike "fxn" as abbreviation. Also, none of the other callbacks have this suffix. None of the callbacks have the namespace prefixed, either, and that doesn't really make sense inside structs (since they have their own scope). How about naming this `on_audio_frame` and renaming (not now) all other callbacks to `on_$event`? There is really no reason to do this rename now though. Feel free to revert to `acb`, since not all of them have been renamed anyway (e.g. `BWController::mcb` is still a TLA).

---

*[toxav/toxav.c, line 78 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT9AbRprr6FjWzO588:-KRT9AbRprr6FjWzO589:b-8if6hy) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/toxav/toxav.c#L78)):*
> ```C
>     toxav_call_cb *toxav_call_fxn;
>     toxav_call_state_cb *toxav_call_state_fxn;
>     toxav_bit_rate_status_cb *toxav_bitrate_status_fxn;
> ```

Why was this moved?

---

*[toxav/toxav.c, line 80 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT8z-ahUwx8c9qcgp3:-KRT8z-boXnRITUIHyRr:b-ylkju) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/toxav/toxav.c#L80)):*
> ```C
>     toxav_bit_rate_status_cb *toxav_bitrate_status_fxn;
> 
>     toxav_audio_receive_frame_cb *toxav_audio_frame_fxn; /* Audio frame receive callback */
> ```

(Awesome comment, I wouldn't have figured out that `audio receive frame cb` means `audio frame receive callback`). Perhaps remove the comment?

---

*[toxcore/Messenger.c, line 1714 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRTB36O_ss7vpBbGZya:-KRTB36O_ss7vpBbGZyb:b6angb8) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/toxcore/Messenger.c#L1714)):*
> ```C
>  */
> void m_callback_msi_packet(Messenger *m, void (*function)(Messenger *m, uint32_t, const uint8_t *, uint16_t, void *,
>                            void *), void *toxav_session)
> ```

Rename to `object` so it's like all the other callbacks that take an extra object. We'll improve that later, but for now it's good to be consistent so you can grep for object.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

JFreegman commented:
*[other/apidsl/tox.in.h, line 1094 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRU2W6PGAXdRnnquLwo:-KRU2W6PGAXdRnnquLwp:bi9g3fy) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/other/apidsl/tox.in.h#L1094)):*
> ```objc
>    * WARNING, calling ${friend.delete} while there's an active ToxAV call will
>    * result in undefined behavior. It's the client's responsibility to end all
>    * ToxAV calls before deleting a friend.
> ```

Why? The correct behaviour would be to have the core automatically end the call before deleting the friend. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

JFreegman commented:
*[other/apidsl/tox.in.h, line 1094 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRU2W6PGAXdRnnquLwo:-KRU6QqK74spFH89gkAT:b-p2gkyg) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/other/apidsl/tox.in.h#L1094)):*
<details><summary><i>Previously, JFreegman wrote…</i></summary>
> Why? The correct behaviour would be to have the core automatically end the call before deleting the friend. 
</details>
Even an error would be better than undefined behaviour. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: 0 of 20 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[other/apidsl/tox.in.h, line 1094 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRU2W6PGAXdRnnquLwo:-KRZkV0KG1eqiXeKW9tB:b-kseqfz) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/other/apidsl/tox.in.h#L1094)):*
<details><summary><i>Previously, JFreegman wrote…</i></summary>
> Even an error would be better than undefined behaviour. 
</details>
We talked about this in IRC, toxcore could error, and not delete the friend. But that would be a change that only exists in this one case, and no one felt good about that. 

ToxAV will delete the session/end the call. But in doing so it'll generate a callback (of which there's no clean way of avoiding) In this case, the callback won't have access to userdata, so userdata will just be `NULL` (and that's the reason for the UB warning).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: 0 of 20 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[other/apidsl/toxav.in.h, line 34 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT7FOEn_7CC8jZVi-i:-KRlJ773PfmUC4uwWRep:b-epualw) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/other/apidsl/toxav.in.h#L34)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Why is this in the public API?
</details>
IMO macros should go into the header. But in this case, I'd forgotten that it's a public API header.

Because it's only used in toxav.c. I've fixed this.

---

*[other/apidsl/toxav.in.h, line 221 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT7NPzgOczW4g0e7eR:-KRlJWGdSdxY0x79VwKf:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/other/apidsl/toxav.in.h#L221)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> `any userdata` like in tox.in.h.
</details>
Done.

---

*[toxav/audio.c, line 99 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT7lEVOj5USXMPbwRq:-KRlJ_p01XGgulc34c5a:b-19zrq1) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/toxav/audio.c#L99)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> I very much dislike "fxn" as abbreviation. Also, none of the other callbacks have this suffix. None of the callbacks have the namespace prefixed, either, and that doesn't really make sense inside structs (since they have their own scope). How about naming this `on_audio_frame` and renaming (not now) all other callbacks to `on_$event`? There is really no reason to do this rename now though. Feel free to revert to `acb`, since not all of them have been renamed anyway (e.g. `BWController::mcb` is still a TLA).
</details>
I'd hurt me to much to use a TLA here. changed to the former.

---

*[toxav/toxav.c, line 78 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT9AbRprr6FjWzO588:-KRlJy6LddkX3cJI9IS7:bmt2sam) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/toxav/toxav.c#L78)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Why was this moved?
</details>
not moved, renamed.

Because I dropped the macro wrapper that also created the userdata (in a struct var called second)

---

*[toxav/toxav.c, line 80 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRT8z-ahUwx8c9qcgp3:-KRlKJEPes29uoG01Z-q:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/toxav/toxav.c#L80)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> (Awesome comment, I wouldn't have figured out that `audio receive frame cb` means `audio frame receive callback`). Perhaps remove the comment?
</details>
Done.

---

*[toxcore/Messenger.c, line 1714 at r1](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRTB36O_ss7vpBbGZya:-KRlKT3WFKExnRAuiWIB:b-r2asop) ([raw file](https://github.com/toktok/toxcore/blob/c98a168c27da86d72800a183e9bc637b2b2837da/toxcore/Messenger.c#L1714)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Rename to `object` so it's like all the other callbacks that take an extra object. We'll improve that later, but for now it's good to be consistent so you can grep for object.
</details>
Done.

But I don't like it...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[toxav/toxav.c, line 78 at r2](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRmWeasGKe637H7rmdh:-KRmWeasGKe637H7rmdi:bfa2ka2) ([raw file](https://github.com/toktok/toxcore/blob/32e7e6b6a6a236b561ddaf0193bc7fd9b38b8ea9/toxav/toxav.c#L78)):*
> ```C
>     pthread_mutex_t mutex[1];
> 
>     toxav_call_cb *toxav_call_fxn;
> ```

Can you rename these things to `on_something` as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

JFreegman commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130#-:-KRoQiEIUrqRI8mXkBOz:bg7igkv)*
<!-- Sent from Reviewable.io -->

JFreegman commented:
Reviewed 12 of 20 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: 1 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[toxav/toxav.c, line 78 at r2](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRmWeasGKe637H7rmdh:-KRpy5w18ZRO-m3R3Qvt:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/32e7e6b6a6a236b561ddaf0193bc7fd9b38b8ea9/toxav/toxav.c#L78)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Can you rename these things to `on_something` as well?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

JFreegman commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 18 of 19 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130#-:-KRq--xyAvx1zeFowpXP:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

nurupo commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 20 files at r1, 17 of 19 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130#-:-KRqQy6837MAyvUWu6CM:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 1 of 19 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[toxav/toxav.c, line 82 at r4](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRrCwG2I9U660idZCVe:-KRrDHvQs6oitgVMVOST:buasmyf) ([raw file](https://github.com/toktok/toxcore/blob/9e134ad3c832abff2831af5cdbb9a444bd703042/toxav/toxav.c#L82)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> This is incorrect. Bit rate status is not bit rate change. It is the "advice" callback.
</details>
I had the same thought. But regardless of what the client does, toxav assumes you'll accept it's bitrate. So the naming here is still consistent. When toxav changes what it want's the bitrate to be. You get this callback, just because the client ignores it, doesn't' change it's name. 

I.e. ignored or not, the next time you get this callback, it'll be lower then the last time. It won't repeat the previous suggestion if ignored. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[toxav/toxav.c, line 82 at r4](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRrCwG2I9U660idZCVe:-KRrHeaA7NZruIQWLeRr:b-1gveh9) ([raw file](https://github.com/toktok/toxcore/blob/9e134ad3c832abff2831af5cdbb9a444bd703042/toxav/toxav.c#L82)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary>
> I had the same thought. But regardless of what the client does, toxav assumes you'll accept it's bitrate. So the naming here is still consistent. When toxav changes what it want's the bitrate to be. You get this callback, just because the client ignores it, doesn't' change it's name. 
> 
> I.e. ignored or not, the next time you get this callback, it'll be lower then the last time. It won't repeat the previous suggestion if ignored. 
> 
> 
</details>
I think it's not about whether toxcore will repeat previous suggestion or not, but about the actual bitrate of the video/audio going through the wire, which toxcore doesn't change on its own.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: 19 of 21 files reviewed at latest revision, 2 unresolved discussions.

---

*[toxav/toxav.c, line 82 at r4](https://reviewable.io:443/reviews/toktok/toxcore/130#-KRrCwG2I9U660idZCVe:-KRrK_3J9WIZRDwQ8y0P:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/9e134ad3c832abff2831af5cdbb9a444bd703042/toxav/toxav.c#L82)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> I think it's not about whether toxcore will repeat previous suggestion or not, but about the actual bitrate of the video/audio going through the wire, which toxcore doesn't change on its own.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

mannol commented on toxav/toxav.c:
```diff
@@ -36,16 +36,18 @@
 #include <stdlib.h>
 #include <string.h>
 
+#define TOXAV_PAIR(TYPE1__, TYPE2__) struct { TYPE1__ first; TYPE2__ second; }
```
why is this prefixed with TOXAV? It's not a part of the public API

mannol commented:
And by change I mean this PR.

iphydf commented:
@mannol: See #40 for the reason for this PR.

---

Review status: 19 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130#-:-KRsG9te1arXx0dkqoxJ:b-nhrijt)*
<!-- Sent from Reviewable.io -->

mannol commented:
This way you loose the ability to provide different user data pointers per callback; I can see that being useful. Anyways, it mostly effects upper levels so this (almost) LGTM.

iphydf commented:
Yes, that is an intentional tradeoff. This way the API is simplified for the common case (most clients either don't use userdata or use a single userdata for all callbacks) and much easier to write FFI (foreign language bindings) for. It also makes it easier to reason about behaviour: currently, any function could potentially invoke a callback. Turns out, indeed some group functions directly invoke callbacks, so not every callback comes from `tox_iterate`. With stateless callbacks, we can be sure that only the iterate functions can ever invoke callbacks.

---

Review status: 19 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130#-:-KRsJe2pXU9zjjItAO9e:b-bfyqqc)*
<!-- Sent from Reviewable.io -->

mannol commented:
Right. Once the ```TOXAV_PAIR``` is renamed I'll lgtm this.

mannol commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130#-:-KRtNSPMYEBHnHe1mJbr:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

mannol commented:
Reviewed 1 of 20 files at r1, 18 of 19 files at r3, 1 of 2 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130)*
<!-- Sent from Reviewable.io -->

iphydf commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 20 files at r1, 18 of 19 files at r3, 1 of 2 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/130#-:-KRtw8ZR9fnqW-K7fp4J:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
unassigned @tux3 and @cleverca22 for being idle. 4/6 approved at time of merge.
